File: StGTest/TestFunctions.cpp
#include "CppUnitTest.h"
#include <opencv2/opencv.hpp>
#include "Common.h"
#include "Functions.h"

#define PASSWORD "People are like water - they will always find a way."

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace StGTest
{
	TEST_CLASS(TestFunctions)
	{
	public:
		TEST_METHOD(TestDetermineEmbeddingMask)
		{
			std::vector<uint8_t> carrierImageBytes = LoadFromFile("..\\..\\files\\input.png");
			cv::Mat carrierImage = Convert::ToCVMat(carrierImageBytes);
			cv::Mat embeddingMask = Functions::DetermineEmbeddingMask(carrierImage);
			std::vector<uint8_t> embeddingMaskBytes = Convert::ToBytes(embeddingMask);
			std::vector<uint8_t> expectedMaskBytes = LoadFromFile("..\\..\\files\\embedding mask.png");
			Assert::IsTrue(expectedMaskBytes == embeddingMaskBytes);
		}

		TEST_METHOD(TestCalculateEffectiveVolume)
		{
			std::vector<uint8_t> carrierImageBytes = LoadFromFile("..\\..\\files\\input.png");
			cv::Mat carrierImage = Convert::ToCVMat(carrierImageBytes);
			cv::Mat embeddingMask = Functions::DetermineEmbeddingMask(carrierImage);
			uint32_t effectiveVolume = Functions::CalculateEffectiveVolume(embeddingMask);
			Assert::IsTrue(effectiveVolume == 469248);
		}

		TEST_METHOD(TestShuffleEmbeddingCoordinates)
		{
			std::vector<uint8_t> carrierImageBytes = LoadFromFile("..\\..\\files\\input.png");
			cv::Mat carrierImage = Convert::ToCVMat(carrierImageBytes);
			cv::Mat embeddingMask = Functions::DetermineEmbeddingMask(carrierImage);
			uint32_t effectiveVolume = Functions::CalculateEffectiveVolume(embeddingMask);
			std::vector<uint8_t> passwordBytes = ToBytes(PASSWORD);
			std::vector<std::tuple<int, int>> shuffledCoordinates = Functions::ShuffleEmbeddingCoordinates(embeddingMask, passwordBytes, effectiveVolume);
			Assert::IsTrue(std::get<0>(shuffledCoordinates[0]) == 593);
			Assert::IsTrue(std::get<1>(shuffledCoordinates[0]) == 459);
		}

		TEST_METHOD(TestEmbed)
		{
			std::vector<uint8_t> carrierImageBytes = LoadFromFile("..\\..\\files\\input.png");
			std::vector<uint8_t> dataBytes = LoadFromFile("..\\..\\files\\message.txt");
			cv::Mat carrierImage = Convert::ToCVMat(carrierImageBytes);
			cv::Mat embeddingMask = Functions::DetermineEmbeddingMask(carrierImage);
			uint32_t effectiveVolume = Functions::CalculateEffectiveVolume(embeddingMask);
			std::vector<uint8_t> passwordBytes = ToBytes(PASSWORD);
			std::vector<std::tuple<int, int>> shuffledCoordinates = Functions::ShuffleEmbeddingCoordinates(embeddingMask, passwordBytes, effectiveVolume);
			cv::Mat stegoImage = Functions::Embed(dataBytes, carrierImage, shuffledCoordinates);
			std::vector<uint8_t> stegoImageBytes = Convert::ToBytes(stegoImage);
			std::vector<uint8_t> expectedBytes = LoadFromFile("..\\..\\files\\output.png");
			Assert::IsTrue(stegoImageBytes == expectedBytes);
		}

		TEST_METHOD(TestExtract)
		{
			std::vector<uint8_t> stegoImageBytes = LoadFromFile("..\\..\\files\\output.png");
			cv::Mat stegoImage = Convert::ToCVMat(stegoImageBytes);
			cv::Mat embeddingMask = Functions::DetermineEmbeddingMask(stegoImage);
			uint32_t effectiveVolume = Functions::CalculateEffectiveVolume(embeddingMask);
			std::vector<uint8_t> passwordBytes = ToBytes(PASSWORD);
			std::vector<std::tuple<int, int>> shuffledCoordinates = Functions::ShuffleEmbeddingCoordinates(embeddingMask, passwordBytes, effectiveVolume);
			std::vector<uint8_t> dataBytes = Functions::Extract(stegoImage, shuffledCoordinates);
			std::vector<uint8_t> expectedBytes = LoadFromFile("..\\..\\files\\received.txt");
			Assert::IsTrue(dataBytes == expectedBytes);
		}
	};
}

File: StGTest/TestFunctions.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using StG;

namespace StGTest
{
    [TestClass]
    public class TestFunctions
    {
        const string DATA_FILE_PATH = "..\\..\\..\\files\\data.txt";
        const string CARRIER_FILE_PATH = "..\\..\\..\\files\\carrier.png";
        const string STEGO_FILE_PATH = "..\\..\\..\\files\\stego.png";

        const int EMBEDDING_BLOCK_SIZE = 2; //2 for small pictures (128x128), 8 for generic photos
        const double EMBEDDING_TRESHOLD = 0.35; //for test purposes only! Use at least 0.5 for practical applications

        const string PASSWORD = "People are like water - they will always find a way.";

        [TestMethod]
        public void TestSanity()
        {
            Assert.IsTrue(true);
        }

        [TestMethod]
        public void TestCompress()
        {
            Byte[] dataBytes = File.ReadAllBytes(DATA_FILE_PATH);
            Byte[] compressedDataBytes = Functions.Compress(dataBytes);
            Assert.IsTrue(compressedDataBytes.Length < dataBytes.Length);
        }

        [TestMethod]
        public void TestDecompress()
        {
            Byte[] dataBytes = File.ReadAllBytes(DATA_FILE_PATH);
            Byte[] compressedDataBytes = Functions.Compress(dataBytes);
            Byte[] resultBytes = Functions.Decompress(compressedDataBytes);
            Assert.AreEqual(dataBytes.Length, resultBytes.Length);
            Random random = new Random();
            int randomIdx = random.Next(0, dataBytes.Length);
            Assert.AreEqual(dataBytes[randomIdx], resultBytes[randomIdx]);
        }

        [TestMethod]
        public void TestEncrypt()
        {
            Byte[] data = File.ReadAllBytes(DATA_FILE_PATH);
            Byte[] password = Encoding.UTF8.GetBytes(PASSWORD);
            AESBridge.Mode mode = AESBridge.Mode.AES256;
            Byte[] encryptedData = Functions.Encrypt(data, password, mode);
            MD5 md5 = MD5.Create();
            string hashData = BitConverter.ToString(md5.ComputeHash(data));
            string hashEncryptedData = BitConverter.ToString(md5.ComputeHash(encryptedData));
            Assert.AreNotEqual(hashData, hashEncryptedData);
        }

        [TestMethod]
        public void TestDecrypt()
        {
            Byte[] data = File.ReadAllBytes(DATA_FILE_PATH);
            Byte[] password = Encoding.UTF8.GetBytes(PASSWORD);
            AESBridge.Mode mode = AESBridge.Mode.AES256;
            Byte[] encryptedData = Functions.Encrypt(data, password, mode);
            Byte[] resultData = Functions.Decrypt(encryptedData, password, mode);
            MD5 md5 = MD5.Create();
            string hashData = BitConverter.ToString(md5.ComputeHash(data));
            string hashResultData = BitConverter.ToString(md5.ComputeHash(resultData));
            Assert.AreEqual(hashData, hashResultData);
        }

        [TestMethod]
        public void TestEmbed()
        {
            Byte[] dataBytes = File.ReadAllBytes(DATA_FILE_PATH);
            Byte[] carrierImageBytes = File.ReadAllBytes(CARRIER_FILE_PATH);
            Byte[] passwordBytes = Encoding.UTF8.GetBytes(PASSWORD);
            Byte[] stegoImageBytes = Functions.Embed(dataBytes, carrierImageBytes, passwordBytes, EMBEDDING_BLOCK_SIZE, EMBEDDING_TRESHOLD);
            Byte[] expectedBytes = File.ReadAllBytes(STEGO_FILE_PATH);
            Assert.AreEqual(expectedBytes.Length, stegoImageBytes.Length);
            Random random = new Random();
            int randomIdx = random.Next(0, expectedBytes.Length);
            Assert.AreEqual(expectedBytes[randomIdx], stegoImageBytes[randomIdx]);
        }

        [TestMethod]
        public void TestExtract()
        {
            Byte[] stegoImageBytes = File.ReadAllBytes(STEGO_FILE_PATH);
            Byte[] passwordBytes = Encoding.UTF8.GetBytes(PASSWORD);
            Byte[] dataBytes = Functions.Extract(stegoImageBytes, passwordBytes, EMBEDDING_BLOCK_SIZE, EMBEDDING_TRESHOLD);
            Byte[] expectedBytes = File.ReadAllBytes(DATA_FILE_PATH);
            Assert.AreEqual(expectedBytes.Length, dataBytes.Length);
            Random random = new Random();
            int randomIdx = random.Next(0, expectedBytes.Length);
            Assert.AreEqual(expectedBytes[randomIdx], dataBytes[randomIdx]);
        }
    }
}
